src/wasm/wasm-binary.cpp

```diff
@@ -4347,6 +4362,103 @@ void WasmBinaryBuilder::visitDrop(Drop* curr) {
   curr->finalize();
 }
 
+void WasmBinaryBuilder::visitTry(Try* curr) {
+  if (debug) {
+    std::cerr << "zz node: Try" << std::endl;
+  }
+  // For simplicity of implementation, like if scopes, we create a hidden block
+  // within each try-body and catch-body, and let branches target those inner
+  // blocks instead.
+  curr->type = getType();
+  curr->body = getBlockOrSingleton(curr->type);
+  if (lastSeparator != BinaryConsts::Catch) {
+    throwError("No catch instruction within a try scope");
+  }
+
+  // Parse instructions within (catch ...)
+  Name label = getNextLabel();
+  breakStack.push_back(
+    {label, curr->type != none && curr->type != unreachable});
+  auto start = expressionStack.size();
+  // catch instruction pushes an exnref value onto the stack, but because we
+  // model catch not as an instruction but a barrier within a try scope for
+  // readability, we assume that after catch there is an exnref value on top of
+  // the stack, and create a pop instruction that pops the exnref value from the
+  // stack. This is only for binaryen's internal representation.
+  Builder builder(wasm);
+  auto* pop = builder.makePop(exnref);
+  expressionStack.push_back(pop);
+  processExpressions();
+  size_t end = expressionStack.size();
+  breakStack.pop_back();
+  auto* block = allocator.alloc<Block>();
+  pushBlockElements(block, start, end);
+  block->name = label;
+  block->finalize(curr->type);
+  curr->catchBody = block;
+  // maybe we don't need a block here?
```

Comment:
this section is confusing to me. can't we use `getBlockOrSingleton()`?

Reply:
I couldn't because I had to insert this code in the middle of `getBlockOrSingleton`, which means, after `catch`, we have to assume that there is an implicit `pop` instruction in Binaryen IR:
```cpp
  // catch instruction pushes an exnref value onto the stack, but because we     
  // model catch not as an instruction but a barrier within a try scope for      
  // readability, we assume that after catch there is an exnref value on top of  
  // the stack, and create a pop instruction that pops the exnref value from the 
  // stack. This is only for binaryen's internal representation.                 
  Builder builder(wasm);                                                         
  auto* pop = builder.makePop(exnref);                                           
  expressionStack.push_back(pop); 
```

I thought about splitting `getBlockOrSingleton` or adding an option to insert some code in the middle, but they seemed not very clean either. Please let me know if you have a better idea.

Reply:
I see, thanks.

Thinking about this, how about adding a parameter to `getBlockOrSingleton` which is how many things can be popped "from the outside"? Then we can add the pop and call that with `1` for that param. And this would maybe prepare us for multivalue in the future where we would use the block type to know how many to pop from outside.

Reply:
Done